File: ecos-records/src/main/java/ru/citeck/ecos/records2/predicate/model/VoidPredicate.java
package ru.citeck.ecos.records2.predicate.model;

import ecos.com.fasterxml.jackson210.annotation.JsonValue;

import java.util.Collections;

public class VoidPredicate implements Predicate {

    public static final VoidPredicate INSTANCE = new VoidPredicate();

    private VoidPredicate() {
    }

    @JsonValue
    protected Object jsonValue() {
        return Collections.emptyMap();
    }

    @Override
    public <T extends Predicate> T copy() {
        @SuppressWarnings("unchecked")
        T result = (T) INSTANCE;
        return result;
    }

    @Override
    public String toString() {
        return "()";
    }

    @Override
    public int hashCode() {
        return VoidPredicate.class.hashCode();
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof VoidPredicate;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records2/predicate/model/VoidPredicate.kt
package ru.citeck.ecos.records2.predicate.model

import com.fasterxml.jackson.annotation.JsonValue

/**
 * This predicate used when predicate is null or not-parsable value.
 * For conditions this predicate should be interpreted as AlwaysTruePredicate
 */
class VoidPredicate private constructor() : Predicate {

    companion object {
        @JvmField
        val INSTANCE = VoidPredicate()
    }

    @JsonValue
    private fun jsonValue(): Any {
        return emptyMap<Any, Any>()
    }

    override fun <T : Predicate> copy(): T {
        @Suppress("UNCHECKED_CAST")
        return INSTANCE as T
    }

    override fun toString(): String {
        return "{}"
    }

    override fun hashCode(): Int {
        return VoidPredicate::class.java.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return other is VoidPredicate
    }
}
